sensing/autoware_pointcloud_preprocessor/include/autoware/pointcloud_preprocessor/outlier_filter/polar_voxel_outlier_filter_node.hpp

```diff
@@ -0,0 +1,194 @@
+  int secondary_noise_threshold_;        // Threshold for primary to secondary return classification
+  std::vector<int64_t> primary_return_types_;    // Return types considered as primary returns
+  std::vector<int64_t> secondary_return_types_;  // Return types considered as secondary returns
+
+  // Diagnostics parameters
+  double visibility_error_threshold_;    // Threshold for visibility diagnostics
+  double visibility_warn_threshold_;     // Warning threshold for visibility diagnostics
+  double filter_ratio_error_threshold_;  // Error threshold for filter ratio diagnostics
+  double filter_ratio_warn_threshold_;   // Warning threshold for filter ratio diagnostics
+
+  // Diagnostics members
+  double visibility_;    // Current visibility value
+  double filter_ratio_;  // Current filter ratio
+  diagnostic_updater::Updater updater_{this};
+  rclcpp::Publisher<autoware_internal_debug_msgs::msg::Float32Stamped>::SharedPtr visibility_pub_;
+  rclcpp::Publisher<autoware_internal_debug_msgs::msg::Float32Stamped>::SharedPtr ratio_pub_;
+  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr noise_cloud_pub_;
+
+  /** \brief Structure to represent a polar voxel index */
+  struct PolarVoxelIndex
+  {
+    int radius_idx;
+    int azimuth_idx;
+    int elevation_idx;
+
+    bool operator<(const PolarVoxelIndex & other) const
+    {
+      if (radius_idx != other.radius_idx) return radius_idx < other.radius_idx;
+      if (azimuth_idx != other.azimuth_idx) return azimuth_idx < other.azimuth_idx;
+      return elevation_idx < other.elevation_idx;
+    }
+  };
+
+  /** \brief Structure to hold both primary and secondary returns for each voxel */
+  struct VoxelPoints
+  {
+    std::vector<size_t> primary_returns;    // Points classified as primary returns
+    std::vector<size_t> secondary_returns;  // Points classified as secondary returns
+  };
+
+  /** \brief Filter using regular PointXYZ (compute polar coordinates)
+   * Uses simple occupancy threshold filtering - voxels with >= voxel_points_threshold points are
+   * kept. No return type classification or ratio-based filtering is applied.
+   */
+  void filter_point_xyz(
+    const PointCloud2ConstPtr & input, const IndicesPtr & indices, PointCloud2 & output);
+
+  /** \brief Filter using PointXYZIRCAEDT (use pre-computed polar coordinates)
+   * Uses advanced two-criteria filtering when return type classification is enabled:
+   * 1. Primary returns >= voxel_points_threshold
+   * 2. Primary-to-secondary ratio >= secondary_noise_threshold
+   * Both criteria must be satisfied for a voxel to be kept.
+   */
+  void filter_point_xyzircaedt(
+    const PointCloud2ConstPtr & input, const IndicesPtr & indices, PointCloud2 & output);
+
+  /** \brief Convert Cartesian coordinates to polar voxel index */
+  PolarVoxelIndex cartesian_to_polar_voxel(double x, double y, double z) const;
+
+  /** \brief Convert Cartesian point to polar coordinates */
+  static void cartesian_to_polar(
```

Comment:
Returning a `std::tuple<double,double,double>` or a custom type `PolarCoordinates{...}` would be more idiomatic than writing to passed-in references.
The return value could be deconstructed at the call site for more readability:
```cpp
auto [radius, azimuth, elevation] = cartesian_to_polar(x, y, z);
```